File: app/src/main/java/com/siju/acexplorer/common/types/FileInfo.java
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.common.types;

import android.os.Parcel;
import android.os.Parcelable;

import com.siju.acexplorer.main.model.groups.Category;


@SuppressWarnings("unused")
public class FileInfo implements Parcelable {

    public static final Creator<FileInfo> CREATOR  = new Creator<FileInfo>() {
        @Override
        public FileInfo createFromParcel(Parcel in) {
            return new FileInfo(in);
        }

        @Override
        public FileInfo[] newArray(int size) {
            return new FileInfo[size];
        }
    };
    private             Category          category;
    private             Category          subcategory;
    private             String            fileName;
    private             String            filePath;
    private             String            fileDate;
    private             String            noOfFilesOrSize;
    private             String            extension;
    private             String            permissions;
    private             String            title;
    private             int               count;
    private             int               icon;
    private             long              id       = -1;
    private             long              bucketId = -1;
    private             long              numTracks;
    private             long              date;
    private             long              size;
    private             long              width;
    private             long              height;

    private             boolean           isDirectory;
    private             boolean           isRootMode;


    // For Images, Audio, Video
    public FileInfo(Category category, long id, long bucketId, String fileName, String filePath,
                    long fileDate, long size,
                    String extension)
    {
        this(category, id, fileName, filePath, fileDate, size, extension);
        this.bucketId = bucketId;
    }


    public static FileInfo createImagePropertiesInfo(Category category, long id, long bucketId, String fileName, String filePath,
                                     long fileDate, long size,
                                     String extension, long width, long height) {
        FileInfo info = new FileInfo(category, id, bucketId, fileName, filePath, fileDate, size, extension);
        info.width = width;
        info.height = height;
        return info;
    }

    public static FileInfo createDummyRecentItem() {
        return new FileInfo(Category.RECENT, 0);
    }


    // Used for apk
    public FileInfo(Category category, long id, String fileName, String filePath, long fileDate,
                    long size,
                    String extension)
    {
        this(category, fileName, filePath, fileDate, size, false, extension, null, false);
        this.id = id;
    }

    public FileInfo(Category category, String fileName, String filePath, long fileDate, long
            noOfFilesOrSize, boolean isDirectory,
                    String extension, String permissions, boolean rootMode)
    {
        this.fileName = fileName;
        this.filePath = filePath;
        this.date = fileDate;
        this.size = noOfFilesOrSize;
        this.isDirectory = isDirectory;
        this.extension = extension;
        this.category = category;
        this.permissions = permissions;
        isRootMode = rootMode;
    }

    // For DialogPicker
    private FileInfo(Category category, String fileName, String filePath, int icon) {
        this.category = category;
        this.fileName = fileName;
        this.filePath = filePath;
        this.icon = icon;
    }

    public static FileInfo createPicker(Category category, String fileName, String filePath, int icon) {
        return new FileInfo(category, fileName, filePath, icon);
    }

    // For HomeLib count
    public FileInfo(Category category, int count) {
        this.category = category;
        this.count = count;
    }

    // For HomeLib count
    public FileInfo(Category category, Category subcategory, int count) {
        this.category = category;
        this.subcategory = subcategory;
        this.count = count;
    }

    //For camera
    public FileInfo(Category category, Category subcategory, String filePath, int count) {
        this.category = category;
        this.subcategory = subcategory;
        this.filePath = filePath;
        this.count = count;
    }

    public static FileInfo createCameraGenericInfo(Category category, Category subcategory, String filePath, int count) {
        return new FileInfo(category, subcategory, filePath, count);
    }

    private FileInfo(Parcel in) {
        category = (Category) in.readSerializable();
        fileName = in.readString();
        filePath = in.readString();
        fileDate = in.readString();
        noOfFilesOrSize = in.readString();
        isDirectory = in.readByte() != 0;
        extension = in.readString();
        id = in.readLong();
        bucketId = in.readLong();
        permissions = in.readString();
        isRootMode = in.readByte() != 0;
        width = in.readLong();
        height = in.readLong();
    }

    public FileInfo(Category category, long albumId, String title, long numTracks) {
        this.category = category;
        this.id = albumId;
        this.title = title;
        this.numTracks = numTracks;
    }

    // Folder Images
    public FileInfo(Category category, long bucketId, String bucketName, String path, int count) {
        this.category = category;
        this.bucketId = bucketId;
        this.fileName = bucketName;
        this.filePath = path;
        this.numTracks = count;

    }

    // App manager
    private FileInfo(Category category, String name, String packageName, long date, long size) {
        this.category = category;
        this.fileName = name;
        this.filePath = packageName;
        this.date = date;
        this.size = size;

    }

    public static FileInfo createAppInfo(Category category, String name, String packageName, long date, long size) {
        return new FileInfo(category, name, packageName, date, size);
    }

    public boolean isRootMode() {
        return isRootMode;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (obj == this) {
            return true;
        }
        if (!(obj instanceof FileInfo)) {
            return false;
        }
        FileInfo o = (FileInfo) obj;
        if (o.filePath == null && this.filePath == null) {
            return true;
        }
        else if (o.filePath == null || this.filePath == null) {
            return false;
        }
        return o.filePath.equals(this.filePath);
    }

    public int getIcon() {
        return icon;
    }

    public void setIcon(int icon) {
        this.icon = icon;
    }

    public Category getCategory() {
        return category;
    }

    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }

    public long getBucketId() {
        return bucketId;
    }

    public void setBucketId(long bucketId) {
        this.bucketId = bucketId;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getExtension() {
        return extension;
    }

    public void setExtension(String extension) {
        this.extension = extension;
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public String getFilePath() {
        return filePath;
    }

    public void setFilePath(String filePath) {
        this.filePath = filePath;
    }

    public String getFileDate() {
        return fileDate;
    }

    public void setFileDate(String fileDate) {
        this.fileDate = fileDate;
    }

    public long getSize() {
        return size;
    }

    public void setSize(long size) {
        this.size = size;
    }

    public long getDate() {
        return date;
    }

    public void setDate(long date) {
        this.date = date;
    }

    public String getNoOfFilesOrSize() {
        return noOfFilesOrSize;
    }

    public void setNoOfFilesOrSize(String noOfFilesOrSize) {
        this.noOfFilesOrSize = noOfFilesOrSize;
    }

    public boolean isDirectory() {
        return isDirectory;
    }

    public void setDirectory(boolean directory) {
        isDirectory = directory;
    }

    public String getPermissions() {
        return permissions;
    }

    public void setPermissions(String permissions) {
        this.permissions = permissions;
    }

    public Category getSubcategory() {
        return subcategory;
    }

    public void setSubcategory(Category subcategory) {
        this.subcategory = subcategory;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public long getNumTracks() {
        return numTracks;
    }

    public void setNumTracks(long numTracks) {
        this.numTracks = numTracks;
    }

    public long getWidth() {
        return width;
    }

    public void setWidth(long width) {
        this.width = width;
    }

    public long getHeight() {
        return height;
    }

    public void setHeight(long height) {
        this.height = height;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeSerializable(category);
        parcel.writeString(fileName);
        parcel.writeString(filePath);
        parcel.writeString(fileDate);
        parcel.writeString(noOfFilesOrSize);
        parcel.writeByte((byte) (isDirectory ? 1 : 0));
        parcel.writeString(extension);
        parcel.writeLong(id);
        parcel.writeLong(bucketId);
        parcel.writeString(permissions);
        parcel.writeByte((byte) (isRootMode ? 1 : 0));
        parcel.writeLong(width);
        parcel.writeLong(height);
    }


}

File: app/src/main/java/com/siju/acexplorer/common/types/FileInfo.kt
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.siju.acexplorer.common.types

import android.os.Parcelable
import com.siju.acexplorer.appmanager.filter.AppSource
import com.siju.acexplorer.main.model.groups.Category
import kotlinx.parcelize.Parcelize

@Parcelize
class FileInfo(var category: Category? = null,
               var subcategory: Category? = null,
               var fileName: String? = null,
               var filePath: String? = null,
               var extension: String? = null,
               var permissions: String? = null,
               var title: String? = null,
               var count : Int = 0,
               var icon : Int = 0,
               var id : Long = -1L,
               var bucketId : Long = -1L,
               var numTracks: Long = 0L,
               var date : Long = 0L,
               var size : Long = 0L,
               var width : Long = 0L,
               var height: Long = 0L,
               var isDirectory: Boolean = false,
               var isRootMode : Boolean = false,
               var source: AppSource = AppSource.PLAYSTORE,
               var systemApp : Boolean = false)  : Parcelable {

    // For Images, Audio, Video
    constructor(category: Category?, id: Long, bucketId: Long, fileName: String?, filePath: String?,
                fileDate: Long, size: Long,
                extension: String?) : this(category, id, fileName, filePath, fileDate, size, extension) {
        this.bucketId = bucketId
    }

    // Used for apk
    constructor(category: Category?, id: Long, fileName: String?, filePath: String?, fileDate: Long,
                size: Long,
                extension: String?) : this(category, fileName, filePath, fileDate, size, false, extension, null, false) {
        this.id = id
    }

    constructor(category: Category?, fileName: String?, filePath: String?, fileDate: Long, noOfFilesOrSize: Long, isDirectory: Boolean,
                extension: String?, permissions: String?, rootMode: Boolean) : this() {
        this.fileName = fileName
        this.filePath = filePath
        date = fileDate
        size = noOfFilesOrSize
        this.isDirectory = isDirectory
        this.extension = extension
        this.category = category
        this.permissions = permissions
        isRootMode = rootMode
    }

    // For DialogPicker
    private constructor(category: Category, fileName: String, filePath: String, icon: Int) : this() {
        this.category = category
        this.fileName = fileName
        this.filePath = filePath
        this.icon = icon
    }

    // For HomeLib count
    constructor(category: Category?, count: Int) : this() {
        this.category = category
        this.count = count
    }

    // For HomeLib count
    constructor(category: Category?, subcategory: Category?, count: Int) : this() {
        this.category = category
        this.subcategory = subcategory
        this.count = count
    }

    //For camera
    constructor(category: Category?, subcategory: Category?, filePath: String?, count: Int) : this() {
        this.category = category
        this.subcategory = subcategory
        this.filePath = filePath
        this.count = count
        this.isDirectory = true
    }

    constructor(category: Category?, albumId: Long, title: String?, numTracks: Long) : this() {
        this.category = category
        id = albumId
        this.title = title
        this.numTracks = numTracks
    }

    // Folder Images
    constructor(category: Category?, bucketId: Long, bucketName: String?, path: String?, count: Int) : this() {
        this.category = category
        this.bucketId = bucketId
        fileName = bucketName
        filePath = path
        numTracks = count.toLong()
        this.isDirectory = true
    }

    // App manager
    private constructor(category: Category, name: String, packageName: String, source: AppSource, isSystemApp: Boolean, date: Long, size: Long) : this() {
        this.category = category
        fileName = name
        filePath = packageName
        this.source = source
        this.systemApp = isSystemApp
        this.date = date
        this.size = size
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }
        if (other === this) {
            return true
        }
        if (other !is FileInfo) {
            return false
        }
        if (other.filePath == null && filePath == null) {
            return true
        } else if (other.filePath == null || filePath == null) {
            return false
        }
        return other.filePath == filePath
    }

    override fun hashCode(): Int {
        var result = category?.hashCode() ?: 0
        result = 31 * result + (subcategory?.hashCode() ?: 0)
        result = 31 * result + (fileName?.hashCode() ?: 0)
        result = 31 * result + (filePath?.hashCode() ?: 0)
        result = 31 * result + (extension?.hashCode() ?: 0)
        result = 31 * result + (permissions?.hashCode() ?: 0)
        result = 31 * result + (title?.hashCode() ?: 0)
        result = 31 * result + count
        result = 31 * result + icon
        result = 31 * result + id.hashCode()
        result = 31 * result + bucketId.hashCode()
        result = 31 * result + numTracks.hashCode()
        result = 31 * result + date.hashCode()
        result = 31 * result + size.hashCode()
        result = 31 * result + width.hashCode()
        result = 31 * result + height.hashCode()
        result = 31 * result + isDirectory.hashCode()
        result = 31 * result + isRootMode.hashCode()
        return result
    }

    companion object {
        fun createImagePropertiesInfo(category: Category?, id: Long, bucketId: Long, fileName: String?, filePath: String?,
                                      fileDate: Long, size: Long,
                                      extension: String?, width: Long, height: Long): FileInfo {
            val info = FileInfo(category, id, bucketId, fileName, filePath, fileDate, size, extension)
            info.width = width
            info.height = height
            return info
        }

        fun createDummyRecentItem(): FileInfo {
            return FileInfo(Category.RECENT, 0)
        }

        fun createPicker(category: Category, fileName: String, filePath: String, icon: Int): FileInfo {
            return FileInfo(category, fileName, filePath, icon)
        }

        fun createCameraGenericInfo(category: Category?, subcategory: Category?, filePath: String?, count: Int): FileInfo {
            return FileInfo(category, subcategory, filePath, count)
        }

        fun createAppInfo(category: Category, name: String, packageName: String, isSystemApp : Boolean, source: AppSource, date: Long, size: Long): FileInfo {
            return FileInfo(category, name, packageName, source, isSystemApp, date, size)
        }
    }
}